Description: Adapt to eclib using INT class for scalar
 In newer eclib versions, scalar is INT (a wrapper around FLINT's fmpz)
 rather than int. Update Cython declarations, constructors, and conversions.
Author: Maximiliano Curia <maxy@debian.org>

Index: sagemath/src/sage/libs/eclib/__init__.pxd
===================================================================
--- sagemath.orig/src/sage/libs/eclib/__init__.pxd
+++ sagemath/src/sage/libs/eclib/__init__.pxd
@@ -10,6 +10,7 @@ from libcpp.map cimport map
 from libcpp.vector cimport vector
 from libcpp.pair cimport pair
 from sage.libs.ntl.types cimport ZZ_c
+from sage.libs.flint.types cimport fmpz_t
 
 
 # NOTE: eclib used to have specific dependencies, so that they had to
@@ -39,7 +40,12 @@ cdef extern from "eclib/newforms.h": pas
 
 # Now the actual declarations, where the order no longer matters
 cdef extern from "eclib/vector.h":
-    ctypedef int scalar
+    cdef cppclass scalar:
+        scalar()
+        scalar(long)
+        scalar(fmpz_t)
+        scalar(const scalar&)
+        void get_fmpz(fmpz_t) const
 
     cdef cppclass vec:
         scalar operator[](long)
@@ -73,8 +79,9 @@ cdef extern from "eclib/smatrix.h":
         scalar elem(int, int) const
         svec row(int) const
 
+ctypedef ZZ_c bigint
+
 cdef extern from "eclib/interface.h":
-    ctypedef ZZ_c bigint
     int I2int(bigint)
 
 cdef extern from "eclib/rat.h":
@@ -119,19 +126,22 @@ cdef extern from "eclib/descent.h":
     cdef cppclass mw:
         mw(Curvedata* curve, int verb, int pp, int maxr)
 
+cdef extern from "eclib/modulus.h":
+    scalar default_modulus[T]()
+
 cdef extern from "eclib/homspace.h":
     cdef cppclass homspace:
-        long modulus
+        long N
         int plusflag # sign
-        int cuspidal
+        long dimension, cuspidal_dimension
 
-        homspace(long n, int hp, int hcusp, int verbose)
+        homspace(long n, scalar mod, int hp, int verbose)
 
-        int h1dim()
-        int h1cuspdim()
-        mat heckeop(long p, int dual, int display)
+        long h1dim()
+        long h1cuspdim()
+        mat heckeop(long p, int cuspidal, int dual, int display)
         vec heckeop_col(long p, int j, int display)
-        smat s_heckeop(long p, int dual, int display)
+        smat s_heckeop(long p, int cuspidal, int dual, int display)
         long h1ncusps()
 
 cdef extern from "eclib/newforms.h":
@@ -143,7 +153,7 @@ cdef extern from "eclib/newforms.h":
         long cuspidal
         long sign
 
-        newforms(long n, int disp)
+        newforms(long n, scalar mod, int disp)
 
         void createfromcurve(int sign, CurveRed CR, int nap)
         void display()
Index: sagemath/src/sage/libs/eclib/homspace.pxd
===================================================================
--- sagemath.orig/src/sage/libs/eclib/homspace.pxd
+++ sagemath/src/sage/libs/eclib/homspace.pxd
@@ -2,3 +2,4 @@ from sage.libs.eclib cimport homspace
 
 cdef class ModularSymbols:
     cdef homspace* H
+    cdef bint _cuspidal
Index: sagemath/src/sage/libs/eclib/homspace.pyx
===================================================================
--- sagemath.orig/src/sage/libs/eclib/homspace.pyx
+++ sagemath/src/sage/libs/eclib/homspace.pyx
@@ -4,8 +4,8 @@ from cysignals.signals cimport sig_on, s
 from cython.operator cimport dereference as deref
 from cython.operator cimport preincrement as inc
 
-from sage.libs.eclib cimport svec, mat, smat
-from sage.libs.eclib.mat cimport MatrixFactory
+from sage.libs.eclib cimport svec, mat, smat, scalar, default_modulus
+from sage.libs.eclib.mat cimport MatrixFactory, scalar_to_integer
 
 from sage.matrix.matrix_space import MatrixSpace
 from sage.rings.integer_ring import ZZ
@@ -49,8 +49,9 @@ cdef class ModularSymbols:
             raise ValueError("sign (= %s) is not supported; use 0, +1 or -1" % sign)
         if level <= 1:
             raise ValueError("the level (= %s) must be at least 2" % level)
+        self._cuspidal = cuspidal
         sig_on()
-        self.H = new homspace(level, sign, cuspidal, verbose)
+        self.H = new homspace(level, default_modulus[scalar](), sign, verbose)
         sig_off()
 
     def __dealloc__(self):
@@ -86,7 +87,7 @@ cdef class ModularSymbols:
             sage: M.level()
             1234
         """
-        return self.H.modulus
+        return self.H.N
 
     #cpdef int dimension(self):
     def dimension(self):
@@ -99,7 +100,7 @@ cdef class ModularSymbols:
             sage: M.dimension()
             156
         """
-        if self.is_cuspidal():
+        if self._cuspidal:
             return self.H.h1cuspdim()
         else:
             return self.H.h1dim()
@@ -151,7 +152,7 @@ cdef class ModularSymbols:
             sage: M = CremonaModularSymbols(1122, cuspidal=True); M.is_cuspidal()
             1
         """
-        return self.H.cuspidal
+        return int(self._cuspidal)
 
     def hecke_matrix(self, long p, dual=False, verbose=False):
         """
@@ -207,7 +208,7 @@ cdef class ModularSymbols:
             True
         """
         sig_on()
-        cdef mat M = self.H.heckeop(p, dual, verbose)
+        cdef mat M = self.H.heckeop(p, int(self._cuspidal), dual, verbose)
         sig_off()
         return MF.new_matrix(M)
 
@@ -271,13 +272,13 @@ cdef class ModularSymbols:
         cdef svec sv
         d = {}
         sig_on()
-        cdef smat M = self.H.s_heckeop(p, dual, verbose)
+        cdef smat M = self.H.s_heckeop(p, int(self._cuspidal), dual, verbose)
         sig_off()
         for i in range(n):
             sv = M.row(i+1)
             iter = sv.begin()
             while iter != sv.end():
-                d[(i,deref(iter).first-1)] = deref(iter).second
+                d[(i,deref(iter).first-1)] = scalar_to_integer(deref(iter).second)
                 inc(iter)
         MS = MatrixSpace(base_ring, n, sparse=True)
         # The next step is the bottleneck.
Index: sagemath/src/sage/libs/eclib/mat.pxd
===================================================================
--- sagemath.orig/src/sage/libs/eclib/mat.pxd
+++ sagemath/src/sage/libs/eclib/mat.pxd
@@ -1,7 +1,31 @@
-from sage.libs.eclib cimport mat
+from sage.libs.eclib cimport mat, scalar
+from sage.rings.integer cimport Integer
+from sage.libs.flint.types cimport fmpz_t
+from sage.libs.flint.fmpz cimport fmpz_init, fmpz_clear, fmpz_get_mpz, fmpz_set_mpz
 
 cdef class Matrix:
     cdef mat* M
 
 cdef class MatrixFactory:
     cdef new_matrix(self, mat M)
+
+cdef inline Integer scalar_to_integer(const scalar& s):
+    cdef fmpz_t z
+    s.get_fmpz(z)
+    cdef Integer res = Integer()
+    fmpz_get_mpz(res.value, z)
+    fmpz_clear(z)
+    return res
+
+cdef inline scalar integer_to_scalar(s):
+    cdef Integer si
+    if isinstance(s, Integer):
+        si = <Integer>s
+    else:
+        si = Integer(s)
+    cdef fmpz_t z
+    fmpz_init(z)
+    fmpz_set_mpz(z, si.value)
+    cdef scalar res = scalar(z)
+    fmpz_clear(z)
+    return res
Index: sagemath/src/sage/libs/eclib/mat.pyx
===================================================================
--- sagemath.orig/src/sage/libs/eclib/mat.pyx
+++ sagemath/src/sage/libs/eclib/mat.pyx
@@ -3,6 +3,7 @@ Cremona matrices
 """
 
 from sage.libs.eclib cimport scalar, addscalar
+from sage.libs.eclib.mat cimport scalar_to_integer, integer_to_scalar
 
 from sage.matrix.matrix_space import MatrixSpace
 from sage.rings.integer_ring import ZZ
@@ -92,7 +93,7 @@ cdef class Matrix:
         if self.M:
             i, j = ij
             if 0<i and i<=self.M[0].nrows() and 0<j and j<=self.M[0].ncols():
-                return self.M.sub(i,j)
+                return scalar_to_integer(self.M.sub(i,j))
             raise IndexError("matrix indices out of range")
         raise IndexError("cannot index into an undefined matrix")
 
@@ -148,7 +149,7 @@ cdef class Matrix:
 ##         """
 ##         return rank(self.M[0])
 
-    def add_scalar(self, scalar s):
+    def add_scalar(self, s):
         """
         Return new matrix obtained by adding `s` to each diagonal entry of ``self``.
 
@@ -162,7 +163,8 @@ cdef class Matrix:
             [3 1]
             [1 2]
         """
-        return new_Matrix(addscalar(self.M[0], s))
+        cdef scalar val = integer_to_scalar(s)
+        return new_Matrix(addscalar(self.M[0], val))
 
     def charpoly(self, var='x'):
         """
@@ -221,7 +223,7 @@ cdef class Matrix:
             Ts = MatrixSpace(ZZ, n, sparse=sparse).zero_matrix().__copy__()
             for i from 0 <= i < n:
                 for j from 0 <= j < n:
-                    Mij = Integer(self.M.sub(i+1,j+1))
+                    Mij = scalar_to_integer(self.M.sub(i+1,j+1))
                     if Mij:
                         Ts.set_unsafe(i, j, Mij)
             return Ts
@@ -229,7 +231,7 @@ cdef class Matrix:
             Td = MatrixSpace(ZZ, n, sparse=sparse).zero_matrix().__copy__()
             for i from 0 <= i < n:
                 for j from 0 <= j < n:
-                    Mij = Integer(self.M.sub(i+1,j+1))
+                    Mij = scalar_to_integer(self.M.sub(i+1,j+1))
                     if Mij:
                         Td.set_unsafe(i, j, Mij)
             return Td
Index: sagemath/src/sage/libs/eclib/newforms.pyx
===================================================================
--- sagemath.orig/src/sage/libs/eclib/newforms.pyx
+++ sagemath/src/sage/libs/eclib/newforms.pyx
@@ -215,7 +215,7 @@ cdef class ECModularSymbol:
             raise ValueError("ECModularSymbol can only be created with signs +1 or 0, not {}".format(sign))
         self.sign = sign
 
-        self.nfs = new newforms(n, 0)
+        self.nfs = new newforms(n, default_modulus[scalar](), 0)
         self.nfs.createfromcurve(sign, CR, nap)
         sig_off()
 
Index: sagemath/src/sage/libs/eclib/wrap.cpp
===================================================================
--- sagemath.orig/src/sage/libs/eclib/wrap.cpp
+++ sagemath/src/sage/libs/eclib/wrap.cpp
@@ -1,6 +1,8 @@
 #include <iostream>
 #include <sstream>
+#include <NTL/ZZ.h>
 using namespace std;
+typedef NTL::ZZ bigint;
 
 /**************** Miscellaneous functions ****************/
 
@@ -128,7 +130,7 @@ int mw_process(struct Curvedata* curve,
                       const bigint* x, const bigint* y,
                       const bigint* z, int sat)
 {
-  Point P(*curve, *x, *y, *z);
+  Point P(curve, *x, *y, *z);
   if (!P.isvalid())
     return 1;
   m->process(P, sat);
